teachertool/src/services/makecodeEditorService.ts

```diff
@@ -3,92 +3,37 @@
-    if (data.type === "pxteditor") {
-        if (data.action === "editorcontentloaded") {
-            readyForMessages = true;
-            sendMessageAsync(); // flush message queue.
-            AutorunService.poke();
-        }
+let driver: IframeDriver | undefined;
 
-        if (data.id && pendingMessages[data.id]) {
-            const pending = pendingMessages[data.id];
-            pending.handler(data);
-            delete pendingMessages[data.id];
-        }
-    }
-}
-
-function sendMessageAsync(message?: any) {
-    return new Promise(resolve => {
-        const sendMessageCore = (message: any) => {
-            logDebug(`Sending message to iframe: ${JSON.stringify(message)}`);
-            makecodeEditorRef!.contentWindow!.postMessage(message, "*");
-        };
-
-        if (message) {
-            message.response = true;
-            message.id = nextId++ + "";
-            pendingMessages[message.id] = {
-                original: message,
-                handler: resolve,
-            };
-            messageQueue.push(message);
-        }
-        if (readyForMessages) {
-            while (messageQueue.length) {
-                sendMessageCore(messageQueue.shift());
-            }
-        }
-    });
-}
+export function setEditorRef(ref: HTMLIFrameElement | undefined) {
+    if (driver) {
+        if (driver.iframe === ref) return;
 
-// Check if the result was successful and (if expected) has data.
-// Logs errors and throws if the result was not successful.
-function validateResponse(result: pxt.editor.EditorMessageResponse, expectResponseData: boolean) {
-    if (!result.success) {
-        throw new Error(`Server returned failed status.`);
+        driver.dispose();
+        driver = undefined;
     }
-    if (expectResponseData && !result?.resp) {
-        throw new Error(`Missing response data.`);
-    }
-}
 
-export function clearReady() {
-    readyForMessages = false;
-}
-
-export function setEditorRef(ref: HTMLIFrameElement | undefined) {
-    makecodeEditorRef = ref ?? undefined;
-    window.removeEventListener("message", onMessageReceived);
     if (ref) {
-        window.addEventListener("message", onMessageReceived);
+        driver = new IframeDriver(ref);
+
+        driver.addEventListener("message", ev => {
+            logDebug(`Message received from iframe: ${JSON.stringify(ev)}`);
+        });
+        driver.addEventListener("sent", ev => {
+            logDebug(`Sent message to iframe: ${JSON.stringify(ev)}`);
+        });
+        driver.addEventListener("editorcontentloaded", ev => {
+            AutorunService.poke();
+        });
     }
 }
 
 //  an example of events that we want to/can send to the editor
 export async function setHighContrastAsync(on: boolean) {
-    const result = await sendMessageAsync({
-        type: "pxteditor",
-        action: "sethighcontrast",
-        on: on,
-    });
-    console.log(result);
+    await driver!.setHighContrast(on)
```

Comment:
Is driver guaranteed to be set here? I can imagine a scenario where we send the teacher tool's high contrast setting to the editor during app load, but the iframe isn't loaded yet.

Reply:
good point. fixed!